primitives/core/src/bls.rs

```diff
@@ -0,0 +1,780 @@
+pub mod bls381 {
+	use crate::crypto::CryptoTypeId;
+	use bls_like::TinyBLS381;
+
+	/// An identifier used to match public keys against BLS12-381 keys
+	pub const CRYPTO_ID: CryptoTypeId = CryptoTypeId(*b"bls8");
+
+	/// BLS12-381 key pair.
+	#[cfg(feature = "full_crypto")]
+	pub type Pair = super::Pair<TinyBLS381>;
+	/// BLS12-381 public key.
+	pub type Public = super::DoublePublicKey<TinyBLS381>;
+	/// BLS12-381 signature.
+	pub type Signature = super::DoubleSignature<TinyBLS381>;
+}
+
+trait BlsBound: EngineBLS + Send + Sync + 'static {}
+
+impl<T: EngineBLS + Send + Sync + 'static> BlsBound for T {}
+
+// Secret key serialized side
+#[cfg(feature = "full_crypto")]
+const SECRET_KEY_SERIALIZED_SIZE: usize =
+	<SecretKey<TinyBLS381> as SerializableToBytes>::SERIALIZED_BYTES_SIZE;
+
+// Public key serialized side
+const PUBLIC_KEY_SERIALIZED_SIZE: usize =
+	<DoublePublicKey<TinyBLS381> as SerializableToBytes>::SERIALIZED_BYTES_SIZE;
+
+// Signature serialized size
+const SIGNATURE_SERIALIZED_SIZE: usize =
+	<DoubleSignature<TinyBLS381> as SerializableToBytes>::SERIALIZED_BYTES_SIZE;
+
+/// A secret seed.
+///
+/// It's not called a "secret key" because ring doesn't expose the secret keys
+/// of the key pair (yeah, dumb); as such we're forced to remember the seed manually if we
+/// will need it later (such as for HDKD).
+#[cfg(feature = "full_crypto")]
+type Seed = [u8; SECRET_KEY_SERIALIZED_SIZE];
+
+/// A public key.
+#[derive(Copy, Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct Public<T> {
+	inner: [u8; PUBLIC_KEY_SERIALIZED_SIZE],
+	_phantom: PhantomData<fn() -> T>,
+}
+
+impl<T> Clone for Public<T> {
+	fn clone(&self) -> Self {
+		Self { inner: self.inner, _phantom: PhantomData }
+	}
+}
+
+impl<T> PartialEq for Public<T> {
+	fn eq(&self, other: &Self) -> bool {
+		self.inner == other.inner
+	}
+}
+
+impl<T> Eq for Public<T> {}
+
+impl<T> PartialOrd for Public<T> {
+	fn partial_cmp(&self, other: &Self) -> Option<sp_std::cmp::Ordering> {
+		self.inner.partial_cmp(&other.inner)
+	}
+}
+
+impl<T> Ord for Public<T> {
+	fn cmp(&self, other: &Self) -> sp_std::cmp::Ordering {
+		self.inner.cmp(&other.inner)
+	}
+}
+
+#[cfg(feature = "full_crypto")]
+impl<T> sp_std::hash::Hash for Public<T> {
+	fn hash<H: sp_std::hash::Hasher>(&self, state: &mut H) {
+		self.inner.hash(state)
+	}
+}
+
+impl<T> Public<T> {
+	/// A new instance from the given PUBLIC_KEY_SERIALIZED_SIZE-byte `data`.
+	///
+	/// NOTE: No checking goes on to ensure this is a real public key. Only use it if
+	/// you are certain that the array actually is a pubkey. GIGO!
+	pub fn from_raw(data: [u8; PUBLIC_KEY_SERIALIZED_SIZE]) -> Self {
+		Public { inner: data, _phantom: PhantomData }
+	}
+
+	/// A new instance from an H1152.
+	///
+	/// NOTE: No checking goes on to ensure this is a real public key. Only use it if
+	/// you are certain that the array actually is a pubkey. GIGO!
+	pub fn from_h1152(x: H1152) -> Self {
```

Comment:
Are these `from_h1152` and `from_h896` really necessary? 
1. I think that their service overlaps the `from_raw`
2. Are used somewhere?
3. Technically they should represent some form of hash output. Here we are using just to hold a bunch of well organized bytes. E.g.. H896 = (48 Bytes for BLS signature + 32*2 for the two scalars for the DLEQ proof).

If the methods are  unused we can remove them otherwise maybe we can instead define simple fixed length integers U896 and U1152 here:
 https://github.com/paritytech/parity-common/blob/master/primitive-types/src/lib.rs ?

Reply:
I have no strong opinion one way or another. I just used ed25519 and ecdsa crypto and I think because they used fixed hash type I did the same. If you prefer to implement it in another way, feel free to do so (and close my pull request) as long as it works with application crypto and hence BEEFY.